desc/integrals/quad_utils.py

```diff
@@ -0,0 +1,247 @@
+        Points to transform.
+    s : float
+        Strength of derivative suppression, s ∈ [0, 1].
+    m : float
+        Number of machine epsilons used for floating point error buffer.
+
+    Returns
+    -------
+    y : jnp.ndarray
+        Transformed points.
+
+    """
+    errorif(not (0 <= s <= 1))
+    # s = 0 -> derivative vanishes like cosine.
+    # s = 1 -> derivative vanishes like cosine^k.
+    y0 = jnp.sin(0.5 * jnp.pi * x)
+    y1 = x + jnp.sin(jnp.pi * x) / jnp.pi  # k = 2
+    y = (1 - s) * y0 + s * y1
+    # y is an expansion, so y(x) > x near x ∈ {−1, 1} and there is a tendency
+    # for floating point error to overshoot the true value.
+    eps = m * jnp.finfo(jnp.array(1.0).dtype).eps
+    return jnp.clip(y, -1 + eps, 1 - eps)
+
+
+def grad_automorphism_sin(x, s=0):
+    """Gradient of sin automorphism."""
+    dy0_dx = 0.5 * jnp.pi * jnp.cos(0.5 * jnp.pi * x)
+    dy1_dx = 1.0 + jnp.cos(jnp.pi * x)
+    dy_dx = (1 - s) * dy0_dx + s * dy1_dx
+    return dy_dx
+
+
+grad_automorphism_sin.__doc__ += "\n" + automorphism_sin.__doc__
+
+
+def tanh_sinh(deg, m=10):
+    """Tanh-Sinh quadrature.
+
+    Returns quadrature points xₖ and weights wₖ for the approximate evaluation of the
+    integral ∫₋₁¹ f(x) dx ≈ ∑ₖ wₖ f(xₖ).
+
+    Parameters
+    ----------
+    deg : int
+        Number of quadrature points.
+    m : float
+        Number of machine epsilons used for floating point error buffer. Larger implies
+        less floating point error, but increases the minimum achievable error.
+
+    Returns
+    -------
+    x, w : (jnp.ndarray, jnp.ndarray)
+        Shape (deg, ).
+        Quadrature points and weights.
+
+    """
+    # buffer to avoid numerical instability
+    x_max = jnp.array(1.0)
+    x_max = x_max - m * jnp.finfo(x_max.dtype).eps
+    t_max = jnp.arcsinh(2 * jnp.arctanh(x_max) / jnp.pi)
+    # maximal-spacing scheme, doi.org/10.48550/arXiv.2007.15057
+    t = jnp.linspace(-t_max, t_max, deg)
+    dt = 2 * t_max / (deg - 1)
+    arg = 0.5 * jnp.pi * jnp.sinh(t)
+    x = jnp.tanh(arg)  # x = g(t)
+    w = 0.5 * jnp.pi * jnp.cosh(t) / jnp.cosh(arg) ** 2 * dt  # w = (dg/dt) dt
+    return x, w
+
+
+def leggauss_lob(deg, interior_only=False):
+    """Lobatto-Gauss-Legendre quadrature.
+
+    Returns quadrature points xₖ and weights wₖ for the approximate evaluation of the
+    integral ∫₋₁¹ f(x) dx ≈ ∑ₖ wₖ f(xₖ).
+
+    Parameters
+    ----------
+    deg : int
+        Number of quadrature points.
+    interior_only : bool
+        Whether to exclude the points and weights at -1 and 1;
+        useful if f(-1) = f(1) = 0. If true, then ``deg`` points are still
+        returned; these are the interior points for lobatto quadrature of ``deg+2``.
+
+    Returns
+    -------
+    x, w : (jnp.ndarray, jnp.ndarray)
+        Shape (deg, ).
+        Quadrature points and weights.
+
+    """
+    N = deg + 2 * bool(interior_only)
+    errorif(N < 2)
+
+    # Golub-Welsh algorithm
+    n = jnp.arange(2, N - 1)
```

Comment:
in case anyone checking my work thoroughly here, and has gotten confused on the indices, here is tip. Recall that after you differentiating the recurrence relation for the legendre polynomial and eventually obtaining the recurrence relation for the derivatives, the resulting set of polynomials you have isn't orthogonal until you throw out the index attached to derivative of the first legendre polynomial.

Reply:
I don't know what this means. Can you explain in more detail?

Reply:
https://www.jstor.org/stable/2004418. At the bottom of page 222, you'll see a matrix multiplication. The indices used there enumerate a set of orthogonal polynomials $P_i$. Here the polynomials we use are derivatives of orthogonal polynomials, call them $Y_i$. If  {$P_1, ... P_n$} is an orthogonal set of polynomials, then  $Y_1, ..., Y_n$ is not, however $Y_2, ... Y_{n+1}$,  so the principal submatrix from page 222 must be used instead of the full matrix

Reply:
Because Y_0 and Y_1 are both constant